### PR TITLE
Trapezoidal scheduler

### PR DESCRIPTION
**Pull request by madragonse**
# Description
First and last step is 0 lr, according to paper, so no training. 
Additional step occurs training in range (n_steps+1) causing a negative lr internally, which is maxed to 0.
Neptune examples of n_steps = 10 with different decay percentages: 

https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9d461d9c-91aa-44f2-858e-4ff023878da8&dash=charts&query=((%60sys%2Ftags%60%3AstringSet%20CONTAINS%20%22trapezoidal_scheduler%22))%20AND%20((%60sys%2Ftags%60%3AstringSet%20NOT%20CONTAINS%20%22test%22))%20AND%20((%60sys%2Ftags%60%3AstringSet%20CONTAINS%20%22remove%22)%20AND%20(%60sys%2Ftags%60%3AstringSet%20CONTAINS%20%22test2%22))&lbViewUnpacked=true&sortBy=%5B%22sys%2Fcreation_time%22%5D&sortFieldType=%5B%22datetime%22%5D&sortFieldAggregationMode=%5B%22auto%22%5D&sortDirection=%5B%22descending%22%5D&compare=IwDg7ArAzANMcPkxQ&compareChartsFilter=lr

image

Comparison to Old-constraints cosine lr scheduler across 25M model grid:
https://app.neptune.ai/o/pmtest/org/llm-random/runs/table?viewId=9d461d9c-91aa-44f2-858e-4ff023878da8&dash=table&compare=IwDg7ArAzANMcPkxQ&compareChartsFilter=lr

On longer training (100 t/ap) is the same:
image

On shorter training  (20 or 40 t/ap) wins (because of a large final_lr_fraction in cosine, and to 0 in trapezoidal):
image

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/check_args.py [hunk 1]
```diff
@@ -38,3 +38,9 @@ def check_args(args):
 
     if args.repeater_mode:
         assert args.relative_init_scale == None
+
+    assert args.lr_trapezoidal_decay_percent + args.lr_warmup_percent <= 1
```
Comment (j321m) [inline research/conditional/utils/check_args.py:42]: does args.lr_warmup_percent always exist?

File: research/conditional/utils/check_args.py [hunk 2]
```diff
@@ -38,3 +38,9 @@ def check_args(args):
...
+    assert (
+        args.lr_trapezoidal_decay_percent + (args.lr_warmup_steps / args.final_lr_step)
+        <= 1
+    )
```
Comment (j321m) [inline research/conditional/utils/check_args.py:46]: there is a function check_args in research/conditional/utils/check_args.py maybe it's a better place for asserts like this one? also final_lr_step can be -1 or None

**Review by j321m — COMMENTED** [review]
nice linear functions in the scheduler, unnecessary asserts that may cause code to crash

File: research/conditional/utils/argparse.py [hunk 3]
```diff
@@ -61,7 +61,8 @@ def introduce_parser_arguments(
     parser.add_argument("--grad_clip", type=float, default=None)
     parser.add_argument("--weight_decay", type=float, default=0.0)
     parser.add_argument("--lr_decay", type=float, default=None)
-    parser.add_argument("--lr_warmup_steps", type=int, default=0)
+    parser.add_argument("--lr_warmup_steps", type=int, default=None)
```
[the remark this answers is not in the document]
Comment (madragonse) [reply research/conditional/utils/argparse.py:64]: idn, changed to 0

File: research/conditional/utils/check_args.py (hunk 1 above)
Comment (madragonse) [reply research/conditional/utils/check_args.py]: changed

File: research/conditional/utils/check_args.py (hunk 2 above)
Comment (madragonse) [reply research/conditional/utils/check_args.py]: changed

**Review by j321m — APPROVED** [review]

**Pull request merged**